File: hanrolink-backend/src/main/java/com/hanrolink/onboarding/service/OnboardingService.java
package com.hanrolink.onboarding.service;

import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.hanrolink.account.entity.BusinessUserAccount;
import com.hanrolink.account.repository.BusinessUserAccountRepository;
import com.hanrolink.business.entity.Business;
import com.hanrolink.business.repository.BusinessRepository;
import com.hanrolink.onboarding.exception.OnboardingAlreadyExistsException;
import com.hanrolink.onboarding.request.OnboardingCreateRequest;
import com.hanrolink.onboarding.response.OnboardingGetResponse;
import com.hanrolink.security.authentication.AuthenticatedUser;
import com.hanrolink.security.authentication.AuthenticatedUserProvider;

@Profile("cognito")
@Service
public class OnboardingService {

  private final AuthenticatedUserProvider authenticatedUserProvider;

  private final BusinessRepository businessRepository;

  private final BusinessUserAccountRepository businessUserAccountRepository;

  public OnboardingService(
    AuthenticatedUserProvider authenticatedUserProvider,
    BusinessRepository businessRepository,
    BusinessUserAccountRepository businessUserAccountRepository
  ) {
    this.authenticatedUserProvider = authenticatedUserProvider;
    this.businessRepository = businessRepository;
    this.businessUserAccountRepository = businessUserAccountRepository;
  }

  /**
   * 初期登録に必要な情報を取得する
   * @param identityProviderSubject 認証プロバイダーのユーザー識別子
   * @param accessToken 認証済みユーザーのアクセストークン
   * @return 初期登録に必要な情報
   */
  public OnboardingGetResponse get(
    String identityProviderSubject,
    String accessToken
  ) {
    AuthenticatedUser currentUser =
      authenticatedUserProvider.get(
        identityProviderSubject,
        accessToken
      );

    return new OnboardingGetResponse(currentUser.email());
  }

  /**
   * 初期登録情報を新規作成する
   * @param identityProviderSubject 認証プロバイダーのユーザー識別子
   * @param accessToken 認証済みユーザーのアクセストークン
   * @param request 初期登録の入力情報
   */
  @Transactional
  public void create(
    String identityProviderSubject,
    String accessToken,
    OnboardingCreateRequest request
  ) {
    if (businessUserAccountRepository.existsByIdentityProviderSubject(identityProviderSubject)) {
      throw new OnboardingAlreadyExistsException();
    }

    AuthenticatedUser currentUser =
      authenticatedUserProvider.get(
        identityProviderSubject,
        accessToken
      );

    Business business = new Business(
      request.business().role(),
      request.business().name(),
      request.business().nameKana(),
      request.business().websiteUrl(),
      request.business().addressPostalCode(),
      request.business().addressPrefecture(),
      request.business().addressMunicipalityStreet(),
      request.business().addressBuilding(),
      request.business().phoneNumber()
    );

    Business savedBusiness = businessRepository.save(business);

    BusinessUserAccount businessUserAccount = new BusinessUserAccount(
      savedBusiness.getId(),
      identityProviderSubject,
      request.businessUserAccount().lastName(),
      request.businessUserAccount().firstName(),
      request.businessUserAccount().lastNameKana(),
      request.businessUserAccount().firstNameKana(),
      request.businessUserAccount().phoneNumber(),
      currentUser.email()
    );

    businessUserAccountRepository.save(businessUserAccount);
  }
}

File: hanrolink-frontend/src/features/onboarding/onboardingService.ts
import type { components } from '../../shared/api/schema'
import { authenticatedApi } from '../../lib/api'
import { getApiErrorMessage } from '../../shared/api/apiError'

type OnboardingInitialData = components['schemas']['OnboardingGetResponse']

type OnboardingSubmission = components['schemas']['OnboardingCreateRequest']

type OnboardingSubmissionResult =
  components['schemas']['OnboardingCreateResponse']

const onboardingFieldLabels: Record<string, string> = {
  'business.role': '事業者区分',
  'business.name': '事業者名',
  'business.nameKana': '事業者名カナ',
  'business.websiteUrl': 'ホームページ',
  'business.addressPostalCode': '郵便番号',
  'business.addressPrefecture': '都道府県',
  'business.addressMunicipalityStreet': '市区町村・番地',
  'business.addressBuilding': '建物名',
  'business.phoneNumber': '事業者電話番号',
  'businessUserAccount.lastName': '担当者の姓',
  'businessUserAccount.firstName': '担当者の名',
  'businessUserAccount.lastNameKana': '担当者の姓カナ',
  'businessUserAccount.firstNameKana': '担当者の名カナ',
  'businessUserAccount.phoneNumber': '担当者電話番号',
}

export async function getOnboardingInitialData(): Promise<OnboardingInitialData> {
  const { data, response } = await authenticatedApi.GET('/api/v1/onboarding')

  if (!response.ok || !data) {
    throw new Error(
      `オンボーディング情報の取得に失敗しました（ステータス:${response.status}）`,
    )
  }
  return data
}

export async function submitOnboarding(
  request: OnboardingSubmission,
): Promise<OnboardingSubmissionResult> {
  const { data, error, response } = await authenticatedApi.POST(
    '/api/v1/onboarding',
    {
      body: request,
    },
  )

  if (!response.ok || !data) {
    throw new Error(
      getApiErrorMessage(
        error,
        `事業者情報の登録に失敗しました。（ステータス：${response.status}）`,
        onboardingFieldLabels,
      ),
    )
  }
  return data
}
